src/persist/src/operators/stream.rs

```diff
@@ -1072,9 +1072,31 @@ mod tests {
             // probe here, because only the conditional input gets sealed. Ideally, we
             // would be able to insert a probe within the conditional_seal operator
             // itself but that's not possible at the moment.
-            for _ in 1..10 {
+            for _ in 0..10 {
                 worker.step();
             }
+
+            // Manually close the inputs in the correct order and give the worker
+            // time to handle each closure individually. We need to do this because
+            // otherwise, both inputs are closed at the exact same time, and the
+            // dataflow is run until all capabilities are dropped.
+            //
+            // There is no guarantee that each operator will observe both operator's
+            // closing at the same time. In particular, the `concat` operator that
+            // introduces the data dependency between the condition input and the
+            // primary input may observe that the primary input has closed before
+            // it observes that the condition input has closed. Then, the input
+            // frontier of the primary seal operator will be 4, and it can inadvertently
+            // seal time 4 (even though we never specifically advanced to time 4).
+            condition_input.close();
+
+            // Give downstream operators a chance to process the condition input
+            // closing.
+            for _ in 0..10 {
```

Comment:
this feels a touch brittle. my gut says there's something better we could do, but I haven't had a chance to load all this up and have a concrete suggestion

on the bright side, it shouldn't actually be "flaky" since step is deterministic /shrug

Reply:
if step/step_or_park was fully deterministic we wouldn't have the nondeterministic bug in the first place :D skimming through the api I think I have a couple of things I could try (e.g. `drop_dataflow` + `step_or_park` until everything closes maybe?) I'll give that a whirl today and see if anything sticks and otherwise ill amend the comment here to be more explicit that this is a brittle choice and merge